devicetypes/qubino/qubino-flush-dimmer.src/qubino-flush-dimmer.groovy

```diff
@@ -0,0 +1,890 @@
+			return integerValue
+	}
+}
+
+private getCommandValue(preference) {
+	def parameterKey = preference.key
+	switch (preference.type) {
+		case "boolean":
+			return settings."$parameterKey" ? preference.optionActive : preference.optionInactive
+		case "boolRange":
+			def parameterKeyBoolean = parameterKey + "Boolean"
+			return !notNullCheck(settings."$parameterKeyBoolean") || settings."$parameterKeyBoolean" ? settings."$parameterKey" : preference.disableValue
+		case "range":
+			return settings."$parameterKey"
+		default:
+			return Integer.parseInt(settings."$parameterKey")
+	}
+}
+
+private notNullCheck(value) {
+	return value != null
+}
+
+private isPreferenceChanged(preference) {
+	if (notNullCheck(settings."$preference.key")) {
+		if (preference.type == "boolRange") {
+			def boolName = preference.key + "Boolean"
+			if (state.currentPreferencesState."$preference.key".status == "disabled") {
+				return settings."$boolName"
+			} else {
+				return state.currentPreferencesState."$preference.key".value != settings."$preference.key" || !settings."$boolName"
+			}
+		} else {
+			return state.currentPreferencesState."$preference.key".value != settings."$preference.key"
+		}
+	} else {
+		return false
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.Command cmd, ep = null) {
+	// Handles all Z-Wave commands we aren't interested in
+	log.debug("zwaveEvent other commands: ${cmd}")
+	[:]
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicReport cmd, ep = null) {
+	log.debug "BasicReport: ${cmd}"
+	dimmerEvents(cmd)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicSet cmd, ep = null) {
+	log.debug "BasicSet: ${cmd}"
+	dimmerEvents(cmd)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelReport cmd, ep = null) {
+	/*
+	This cmd value reports inaccurate values which often differ from the values reported by SwitchMultilevelSet by 1-3%.
+	Probbably we should only depend on SwitchMultilevelSet report.
+	*/
+	log.debug "SwitchMultilevelReport: ${cmd}"
+	//dimmerEvents(cmd)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelSet cmd, ep = null) {
+	log.debug "SwitchMultilevelSet: ${cmd}"
+	dimmerEvents(cmd)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelStartLevelChange cmd) {
+	log.debug "SwitchMultilevelStartLevelChange: ${cmd}"
+	def result = []
+	result << createEvent(name: "switch", value: "on")
+	result << createEvent(name: "level", value: cmd.startLevel)
+	return result
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelStopLevelChange cmd) {
+	log.debug "SwitchMultilevelStopLevelChange: ${cmd}"
+	refresh()
+}
+
+private dimmerEvents(physicalgraph.zwave.Command cmd, ep = null) {
+	def value = (cmd.value ? "on" : "off")
+	def result = [createEvent(name: "switch", value: value)]
+	if (cmd.value <= 100) {
+		result << createEvent(name: "level", value: cmd.value == 99 ? 100 : cmd.value)

```

Comment:
I'd recommend switching from `list << object` to `list += object`

Reply:
@PKacprowiczS "<<" may be tricky when it comes to adding a list to a list. In case of adding a map to a list of maps it should be fine (I verified that).

Reply:
It's even better when adding list to a list. Adding with `<<` operator results with nested list, which later cause some problems with returning events from _parse()_ method.

Reply:
@PKacprowiczS that's right (nested lists), but we have a map being added to a list here.

Reply:
Afaik event is an object, so in this particular case we're adding an object to a list.

Reply:
@PKacprowiczS Then the problem with nested lists doesn't apply here.

Reply:
So I guess we have a little misunderstaning here. I'm not saying that there's something wrong in this particular use of `<<` operator, it was just my recommendation to get used to using `+=` over `<<` to not have such problems with nested lists in the future (which, for a newcomer like Bartek, might take some time to notice by himself)

Reply:
Ok, thank you. So I will leave it like that here and just remember to use "+=" with nested lists in the future. I think the "<<" may give better performance as it adds the elements to the existing list instead of creating new list every time, right?